url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/io/image/TargaImage.java

```diff
@@ -0,0 +1,345 @@
+//===========================================================================
+package vsdk.toolkit.io.image;
+
+import java.io.BufferedInputStream;
+import java.io.DataInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+
+import java.awt.Dimension;
+import java.awt.image.DirectColorModel;
+
+import vsdk.toolkit.io.PersistenceElement;
+import vsdk.toolkit.media.RGBImage;
+import vsdk.toolkit.media.RGBAImage;
+
+/**
+The TargaImage class represents a targa image (the image files with .tga
+extension)
+*/
+
+public class TargaImage extends PersistenceElement
+{
+    private static final int NO_TRANSPARENCY = 255;
+    private static final int FULL_TRANSPARENCY = 0;
+    
+    private short idLength;
+    private short colorMapType;
+    private short imageType;
+    private int cMapStart;
+    private int cMapLength;
+    private short cMapDepth;
+    private int xOffset;
+    private int yOffset;
+    private int xSize;
+    private int ySize;
+    private short pixelDepth;
+    private short imageDescriptor;
+    private DirectColorModel cm;
+    public byte[] pixels;
+
+    /**
+    Constructs a TGAImage given the File image
+    @param srcFile The tga image file
+    */
+    public TargaImage(File srcFile) throws ImageNotRecognizedException {
+        try
+        {
+            open(srcFile);
+        }
+        catch(IOException ioe)
+        {
+            throw new ImageNotRecognizedException("I/O Error while reading the image", srcFile);
+        }
+    }
+    
+    private void open(File srcFile) throws ImageNotRecognizedException, IOException {
+        byte red = 0;
+        byte green = 0;
+        byte blue = 0;
+        byte alpha = FULL_TRANSPARENCY;
+        int srcLine = 0;
+        
+        FileInputStream fis = new FileInputStream(srcFile);
+        BufferedInputStream bis = new BufferedInputStream(fis, 8192);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *bis*. The resource is referenced at lines 65, 66, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 95, 98, 115, 116, 117, 119, 148, 152, 153, 154, 157, 181, 182, 183, 185. The resource is closed at line 212. There are other execution paths that do not contain closure statements, for example, when an else condition at lines 80, 86, 97 evaluate to true. Ensure *bis* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).